design/design-doc-composition-functions.md

```diff
@@ -209,570 +738,353 @@ metadata:
-desired:
-  resources:
-  - name: cloudsqlinstance
-    resource:
-      apiVersion: database.gcp.crossplane.io/v1beta1
-      kind: CloudSQLInstance
-      spec:
-        forProvider:
-          databaseVersion: POSTGRES_9_6
-          region: us-central1
-          settings:
-            tier: db-custom-1-3840
-            dataDiskType: PD_SSD
-            dataDiskSizeGb: 20
-        writeConnectionSecretToRef:
-          namespace: crossplane-system
-          name: cloudsqlpostgresql-conn
-    connectionDetails:
-    - name: hostname
-      fromConnectionSecretKey: hostname
-    readinessChecks:
-    - type: None
+Starlark is a (very) limited dialect of Python designed for configuration. It
+embeds into many languages, including Go. This makes it a great candidate to
+build a generic Function that allows Composition authors to provide _inline_
+logic in a general-purpose-ish programming language. This is similar to the
+["GitHub Script"][github-script] GitHub Action that lets you plug some arbitrary
+logic into a GitHub Action pipeline when no existing Action does quite what you
+need.
+
+#### Iterating on Compositions
+
+Getting fast feedback on a Crossplane Composition has historically been a pain
+because the Composition logic has been encapsulated in a Kubernetes controller.
+In order to know whether your Composition works as intended you need to:
+
+1. Deploy (the right version of!) Crossplane to a Kubernetes cluster.
+2. Install Providers and Configurations for any resources you want to compose.
+3. Apply your XRD and Composition (potentially via a Configuration).
+4. Create an XR (or claim).
+5. See whether it works.
+
+Moving Composition logic out of Crossplane and into versioned Functions makes it
+a lot easier to test and iterate on, client-side. For example a `function test`
+command could test a single Function in isolation:
+
+```shell
+# Test a single Function by passing it a JSON RunFunctionRequest. This pulls and
+# starts the Function, makes the request, stops it, then returns the result.
+$ kubectl crossplane function test xpkg.upbound.io/negz/go-templates:v0.1.0 run-function-request.json
+{
+  # JSON encoded RunFunctionResponse omitted for brevity.
+}
 ```
 
-A function that wanted to set only an XR connection detail could return:
+You could similarly imagine a more holistic, `helm template`-like experience
+that locally rendered an entire Composition by pulling and running its templates
+locally:
 
-```yaml
-apiVersion: apiextensions.crossplane.io/v1alpha1
-kind: FunctionIO
-observed: {}  # Omitted for brevity.
-desired:
-  composite:
-    connectionDetails:
-    - type: FromValue
-      name: username
-      value: admin
+```shell
+$ kubectl crossplane composition render composition.yaml xr.yaml
+---
+# YAML stream omitted for brevity.
 ```
 
-A function wishing to delete a composed resource may do so by setting its
-`resource` to null, for example:
+Assuming a 'pure' Composition consisting only of Functions with no P&T
```

Comment:
> Assuming a 'pure' Composition consisting only of Functions with no P&T resources, 

Could we end up with P&T transform logic becoming yet another Function at some point and `spec.resources` disappearing with v2 of the Compositions :)? In this case, we wouldn't need assumptions like this and provide a more wholistic experience.

Reply:
Nvm, just saw the `Patch-and-Transform as a Function` section :)

Reply:
i think this answers my questions on how can we speed up the development process and allow folks to run their composition "inner loop" more in isolation when they do have patch & transform logic also...

to confirm: it would not be possible to use something like `composition render` unless P&T is also supported/converted to a Function also? 🤔

Reply:
> to confirm: it would not be possible to use something like composition render unless P&T is also supported/converted to a Function also?

@jbw976 I think it would be possible, but messy.

If all of a Composition's logic is packaged as versioned artifacts (i.e. Functions) then a client (e.g. `composition render`) can easily pull down and test against the same set of artifacts that will be used in production.

If some of the logic is baked into Crossplane then presumably we need to bake that logic into the client (`composition render`) too. Then it becomes important that the client is running the same code as the server (i.e. Crossplane).

It's not an unsolvable problem - maybe you just make sure to use `kubectl crossplane` v1.14 if you're using Crossplane v1.14 in production. Maybe `kubectl crossplane composition render` spins up a `kind` cluster and actually installs a specific version of Crossplane to test your composition. It's just a much easier problem to solve when everything is a Function.